src/riverroute/mosart_budget_type.F90

```diff
@@ -0,0 +1,261 @@
+            end if
+            ! this was for budget_terms(17)
+            !if (nt==1) then
+            !  this%in_grc(nr,nt)=this%in_grc(nr,nt) +ctl%qirrig(nr)
+            !endif
+         end do
+      end do
+
+      this%end_vol_grc(:,:) = 0.0_r8
+      this%out_grc(:,:) = 0.0_r8
+      this%net_grc(:,:) = 0.0_r8
+      this%lag_grc(:,:) = 0.0_r8
+
+      this%beg_vol_glob(:) = 0.0_r8
+      this%end_vol_glob(:) = 0.0_r8
+      this%in_glob(:) = 0.0_r8
+      this%out_glob(:) = 0.0_r8
+      this%net_glob(:) = 0.0_r8
+      this%lag_glob(:) = 0.0_r8
+
+   end subroutine set_budget
+
+   !-----------------------------------------------------------------------
+
+   subroutine check_budget(this, begr, endr, ntracers, dt)
+
+      ! Arguments
+      class(budget_type)   :: this
+      integer, intent(in)  :: begr, endr, ntracers
+      real(r8), intent(in) :: dt
+
+      ! Local variables
+      integer  :: nr, nt                !indecies
+      integer  :: nt_liq                ! liquid index
+      integer  :: yr,mon,day,ymd,tod    !time vars
+      real(r8) :: tmp_in(6, ntracers)   ! array to pass to mpi_sum
+      real(r8) :: tmp_glob(6, ntracers) ! array from mpi_sum
+      logical  :: error_budget          ! flag for an error
+      real(r8) :: abserr, relerr
+      !-------------------------------------------------
+
+      call get_curr_date(yr, mon, day, tod)
+      ymd = yr*10000 + mon*100 + day
+      tmp_in = 0.0_r8
+      tmp_glob = 0.0_r8
+
+      nt_liq = ctl%nt_liq
+      do nr = begr, endr
+         do nt = 1, ntracers
+            this%end_vol_grc(nr, nt) = ctl%volr(nr, nt)
+            this%out_grc(nr, nt) = this%out_grc(nr, nt) + ctl%direct(nr, nt) + ctl%direct_glc(nr, nt)
+            if (nt == nt_liq) then
+               this%out_grc(nr, nt) = this%out_grc(nr, nt) + ctl%flood(nr)
+            end if
+            if (ctl%mask(nr) >= 2) then
+               this%out_grc(nr, nt) = this%out_grc(nr, nt) + ctl%runoff(nr, nt)
+            else
+               this%lag_grc(nr, nt) = this%lag_grc(nr, nt) - ctl%erout_prev(nr, nt) - ctl%flow(nr, nt)
+            end if
+            this%out_grc(nr,nt) = this%out_grc(nr,nt) * dt
+            this%lag_grc(nr,nt) = this%lag_grc(nr,nt) * dt
+            this%net_grc(nr,nt) = this%end_vol_grc(nr,nt) - this%beg_vol_grc(nr,nt) - (this%in_grc(nr,nt)-this%out_grc(nr,nt))
+            this%accum_grc(nr,nt) = this%accum_grc(nr,nt) + this%net_grc(nr,nt)
+         end do
+      end do
+
+      do nt = 1, ntracers
+         tmp_in(1, nt) = sum(this%beg_vol_grc(:, nt))
+         tmp_in(2, nt) = sum(this%end_vol_grc(:, nt))
+         tmp_in(3, nt) = sum(this%in_grc(:, nt))
+         tmp_in(4, nt) = sum(this%out_grc(:, nt))
+         tmp_in(5, nt) = sum(this%net_grc(:, nt))
+         tmp_in(6, nt) = sum(this%lag_grc(:, nt))
+      end do
+
+      tmp_in = tmp_in*1e-6_r8 !convert to million m3
+      call shr_mpi_sum(tmp_in, tmp_glob, mpicom_rof, 'mosart global budget', all=.false.)
```

Comment:
However, ESMF_VMAllReduce only accepts one-dimensional arrays. tmp_in and tmp_glob are 2d-arrays. So copies will need to be made and 6 calls will need to be done and then copies need to be made. Frankly, I am worried that moving to only ESMF MPI communication will make the code harder to understand since it really supports only a very small subset of MPI communications. And I think it is not a good idea to mix ESMF calls in the code with MPI calls. Happy to talk about this some more.

Reply:
See issue https://github.com/ESCOMP/MOSART/issues/100.